### PR TITLE
Add a Script extension that works with Director

### PR DESCRIPTION
**Pull request by ysims**
Allows us to do

```cpp
emit<Script>(std::make_unique<LimbsSequence>(), "KickRight.yaml");
```

Then that'll load the script and emit a LimbsSequence for that script.

Future work: cache scripts

### CODE REVIEW

File: shared/extension/NewScript.hpp [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+            Target(Target&& other) noexcept
+                : id(other.id), position(other.position), gain(other.gain), torque(other.torque) {}
+            Target& operator=(const Target& other) = default;
+            Target& operator                       =(Target&& other) noexcept {
+                id       = other.id;
+                position = other.position;
+                gain     = other.gain;
+                torque   = other.torque;
+                return *this;
+            }
```
Comment (CMurtagh-LGTM) [inline shared/extension/NewScript.hpp:41]:
```suggestion
```
You use initialiser lists.

File: shared/extension/NewScript.hpp [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+                                    std::system_category(),
+                                    ("Failed to extract platform name from '" + hostname + "'."));
+        }
+
+        /// @brief If a robot-specific script exists, load the YAML file as a script. If not, loads the
+        /// platform-specific script. If neither exist, throws a runtime error.
+        /// @param script The name of the script to run.
+        /// @return The script contents in the form off a YAML node
+        /// @throws Runtime error when the script doesn't exist.
+        static YAML::Node load_script(const std::string& script) {
```
Comment (CMurtagh-LGTM) [inline shared/extension/NewScript.hpp:96]:
```suggestion
        static YAML::Node load(const std::string& script) {
```
Doubling up on the name script, `Script::load_script` vs `Script::load`.

File: shared/extension/NewScript.hpp [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+            float position{0.0f};
+            /// @brief Gain for this servo in this frame
+            float gain{0.0f};
+            /// @brief Torque for this servo in this frame
+            float torque{0.0f};
+        };
+
+        Frame() : duration() {}
+        Frame(const NUClear::clock::duration& dur, std::vector<Target> targets)
+            : duration(dur), targets(std::move(targets)) {}
```
Comment (CMurtagh-LGTM) [inline shared/extension/NewScript.hpp:55]:
```suggestion
```
There's no point here as well.

File: shared/extension/NewScript.hpp [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+
+        /// @brief The time it should take for the servos to move to their targets
+        NUClear::clock::duration duration;
```
Comment (CMurtagh-LGTM) [inline shared/extension/NewScript.hpp:58]:
```suggestion
        NUClear::clock::duration duration{};
```
Initialise it, if you delete the constructor.

**Review by CMurtagh-LGTM — CHANGES_REQUESTED** [review]
Please don't call it `NewScript`.

**Comment by ysims** [comment]
> Please don't call it `NewScript`.

Do you have a suggestion? :P maybe we need a director branch already after all :weary:

File: shared/extension/behaviour/Script.hpp [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+
+    template <typename Sequence>
+    struct Script {
+        std::vector<Frame> frames;
+
+        /// @brief Assumes the hostname is of the form <platform_name><robot_number> and returns <platform_name>.
+        /// @param hostname The hostname of the device the program is running on.
+        /// @return The platform name.
+        /// @throws System error when the platform cannot be extracted from the hostname.
+        static inline std::string getPlatform(const std::string& hostname) {
```
Comment (TrentHouliston) [inline shared/extension/behaviour/Script.hpp:54]:
```suggestion
        static inline std::string get_platform(const std::string& hostname) {
```

**Review by TrentHouliston — COMMENTED** [review]
One minor point

**Review by TrentHouliston — APPROVED** [review]

**Comment by CMurtagh-LGTM** [comment]
> because he's busy teaching johanne maths
:stuck_out_tongue:

**Pull request merged**